pinot-core/src/main/java/org/apache/pinot/core/indexsegment/mutable/MutableSegmentImpl.java

```diff
@@ -204,20 +205,22 @@ public boolean index(GenericRow row) {
     // If metrics aggregation is enabled and if the dimension values were already seen, this will return existing docId,
     // else this will return a new docId.
     int docId = getOrCreateDocId(dictIdMap);
-
+    boolean canTakeMore = false;
     // docId == numDocs implies new docId.
     if (docId == numDocs) {
       // Add forward and inverted indices for new document.
       addForwardIndex(row, docId, dictIdMap);
       addInvertedIndex(docId, dictIdMap);
       // Update number of document indexed at last to make the latest record queryable
-      return _numDocsIndexed++ < _capacity;
+      canTakeMore = _numDocsIndexed++ < _capacity;
     } else {
-      Preconditions
-          .checkState(_aggregateMetrics, "Invalid document-id during indexing: " + docId + " expected: " + numDocs);
+      Preconditions.checkState(_aggregateMetrics, "Invalid document-id during indexing: " + docId + " expected: " + numDocs);
       // Update metrics for existing document.
-      return aggregateMetrics(row, docId);
+      canTakeMore = aggregateMetrics(row, docId);
     }
+    // update indexing time
+    _lastIndexedTimestamp = System.currentTimeMillis();
```

Comment:
If there is a delay in the stream, that will show up as a pinot delay. Another choice is to take the last _polled_ timestamp. That will ensure that as long as Pinot polls the stream, the available data is considered fresh. Not sure why we are not doing that.  A stream might (and probably should) have its own staleness metric being tracked and monitored so that an application does not need to track a staleness metric via pinot and root cause it to the stream if that is indeed the case.

If you want to continue with this approach, then I suggest that you extend the interface to retrieve rows from a stream to include the highest available timestamp of a record _in_ the stream. That way, streams that provide the staleness metric can translate (via Pinot) to the staleness monitoring application.

Reply:
The user defined SLA will account for any expected delays in the incoming stream itself. Just going by the poll time can miss cases where we have issues with indexing (there have been a few in the past; for ex, issues with aggregatable-metrics for multi-value columns etc).

The timestamp in the record can have a very coarse granularity, for ex, days - which isn't ideal for the granularity we would like to track freshness at. Also, see motivation notes about the lack of other usable timestamps.

Reply:
Issues with indexing the row should be dealt with as a bug and fixed, since that is a loss of data. 

Also, I was proposing to extend the interface so that a stream that keeps track of timestamps of incoming messages (for its own SLA) can use the mechanism. I was not referring to the time column of the schema

We can chat a little to see if any of these ideas can be explored further before discarding